protocols/gossipsub/src/handler.rs

```diff
@@ -92,12 +92,16 @@ pub struct GossipsubHandler {
     peer_kind: Option<PeerKind>,
 
     /// Keeps track on whether we have sent the peer kind to the behaviour.
+    //
     // NOTE: Use this flag rather than checking the substream count each poll.
     peer_kind_sent: bool,
 
     /// If the peer doesn't support the gossipsub protocol we do not immediately disconnect.
     /// Rather, we disable the handler and prevent any incoming or outgoing substreams from being
     /// established.
+    //
+    // mxinden: Why do this?
```

Comment:
IIRC the default behaviour for an unsupported protocol is for the handler to throw an error. 

Throwing an error there means an instant disconnection. For a handler that is composed of multiple protocols, it may be useful to allow peers to support some protocols and not others. In this case if gossipsub is not supported, the peer is not immediately disconnected and libp2p can continue to function on other protocols. 

In the case that there is only gossipsub, this will not create any substreams, set the KeepAlive to False and the peer will be disconnected (as typical composed handlers compose the KeepAlive parameter)

I could be missing something with the design, but I think this makes it generic for multiple protocols, such that we don't force all peers to support gossipsub if gossipsub is supported locally.